[torch.special] Adding betainc, betaincc, betaincinv, betainccinv, betaln and beta with backward operation

voidbag opened the pull request:
Fixes #132133

#58700 and #129836 don't support GPU, due to 'gpu_kernel_with_scalars' with only one scalar value.
I changed it to 'gpu_kernel_with_scalars_tenary' which is new function I made.

The contribution point in #129836 except re-indentation is 
```
#include <ATen/ops/special_betainc.h>
#include <ATen/ops/special_betainc_native.h>
```
I made this part Co-authored-by @richwomanbtc

I implemented special.betainc (cpu and gpu) based on #58700, #129836
I implemented betainc backward operation, based on "https://github.com/tensorflow/probability/blob/b14ae1d79de4a52d834a3ba2dc88f7e5d849e6c7/tensorflow_probability/python/math/special.py#L432-L491"
I implemented forward/backward operation of betaln, betaincinv based on
"https://github.com/tensorflow/probability/blob/751461f84b8d29c83216d719950b8d999c7381d5/tensorflow_probability/python/math/special.py#L1854-L2066" and
"https://github.com/tensorflow/probability/blob/751461f84b8d29c83216d719950b8d999c7381d5/tensorflow_probability/python/math/special.py#L567-L868", respectively.



Things to consider
1. Co-Authoring method
    - first and second commits are followed by Co-authored-by name <email>
    - I'm now sure if this way is correct.

2. let me know if your name or email has to be changed.
    - @nicola-decao: Nicola De Cao <nicola.decao@gmail.com>
      (first commit: https://github.com/pytorch/pytorch/pull/132135/commits/bd14d16a722dfa505a4c24ec2c83a859b6c157d8)
    - @richwomanbtc: richwomanbtc <richwomanbtc@gmail.com>
      (second commit: https://github.com/pytorch/pytorch/pull/132135/commits/336a0471d6961c89c962c82e7fb0524837abd080)

3. Does fourth commit (add 'gpu_kernel_with_scalars_ternary') need to be separate PR?
    - gpu_kernel_with_scalars_ternary function was added into "/aten/src/ATen/native/cuda/Loops.cuh"
    - ```int max_cpu_scalars_on_non_cpu = config.allow_cpu_scalars_ ? 1 : 0```
      -> ```int max_cpu_scalars_on_non_cpu = config.allow_cpu_scalars_ ? config.num_inputs_ - 1 : 0```

4. Inefficient Indefinite loop(control flow)? in "intorch/csrc/autograd/FunctionsManual.cpp"
    - Better than ```stop.all().item<bool>()```? (ex. tf.while_loop)

5. The betainc fwd implementation has been changed to Eigen's implementation like tensorflow_probability
    - Is it Okay?


Usage of Docker Image(betainc-bwd): 
- You can use this feature with this command before #132135  is merged.
  ~`docker pull voidbag/pytorch:2.6.0a0-cuda12.1-cudnn9-devel-wip-betainc-with-bwd-voidbag-v1.1.0`~
  ~`docker pull voidbag/pytorch:2.6.0a0-cuda12.1-cudnn9-devel-wip-betainc-with-bwd-voidbag-v1.2.0`~
  ~(v1.2.0 supports zero-dimensional tensors)~
  ~`docker pull voidbag/pytorch:2.6.0a0-cuda12.1-cudnn9-devel-wip-betainc-with-bwd-voidbag-v1.3.0`~
  ~(v1.3.0 supports forward-mode AD and it fixed torch.special.betaincinv's bug in fwd and bwd)~
  `docker pull voidbag/pytorch:2.7.0a0-cuda12.6-cudnn9-devel-wip-betainc-with-bwd-voidbag-v1.4.0`
  (v1.4.0 make the function signiture of beta family same as scipy.special.(betainc|betaincc|betaincinv|betainccinv).
  **The parameter orders of random variable X and Y are moved from first parameter to last parameter!**
  ex) `betainc(x, a, b) -> betainc(a, b, x), betaincinv(y, a, b) -> betaincinv(a, b, y)`
  
  (blackwell support: `docker pull voidbag/pytorch:2.8.0a0-cuda12.8-cudnn9-devel-wip-betainc-with-bwd-voidbag-v1.4.0`)
  [Docker Hub Repository Link: voidbag/pytorch](https://hub.docker.com/r/voidbag/pytorch)



cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @albanD @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @azahed98 @H-Huang @chenyang78 @kwen2501 @c-p-i-o @yf225

voidbag committed:
[torch.special] Adding betainc

PR: #58700 has been implemented in recent repo.
(https://github.com/pytorch/pytorch/pull/58700)
It cannot be compiled yet.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>
Co-authored-by: Nicola De Cao <nicola.decao@gmail.com>

voidbag committed:
Adding betainc

The PR #129836 has been implemented.
(https://github.com/pytorch/pytorch/pull/129836)

CPU version in this version can be compiled.

The line 'build_ternary_op' in BinaryOps.cpp has been removed, for
scalar parameter.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>
Co-authored-by: richwomanbtc <richwomanbtc@gmail.com>

voidbag committed:
gpu_kernel_with_scalar to gpu_kernel

gpu_kernel_with_scalars can handle at most 2 parameters.
But, betainc requires 3 parameters.

To compile source code, gpu_kernel_with_scalars needs to be changed to
gpu_kernel.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
betainc impl has been changed that of eigen repo

betainc implementation has been changed from custom impl to eigen impl
used in tensorflow_probability.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
add 'gpu_kernel_with_scalars_ternary'

Current 'gpu_kernel_with_scalars' only supports 2 parameters.
But, some probability distributions requires 1 random variable and two
parameters. like gaussian(x, mu, sigma), beta(x, alpha, beta), etc.
This 'gpu_kernel_with_scalars_ternary' can support distributions with two
parameters.

To support 2 scalars, the maximum number of allowed scalars has to be
set as num_inputs_ - 1, not 1(current) in TensorIterator.cpp

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
special.betainc supports 2 scalar parameters

special.betainc supports 2 scalar parameters.
(x: Tensor, alpha: Scalar, beta: Scalar)

It doesn't support following 2 cases.
(x: Scalar, alpha: Tensor, beta: Scalar)
(x: Scalar, alpha: Scalar, beta: Tensor)

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
special.betainc: support backward operation

It has been implemented based on tensorflow_probability impl

- Reference: https://github.com/tensorflow/probability/blob/
  b14ae1d79de4a52d834a3ba2dc88f7e5d849e6c7/
  tensorflow_probability/python/math/special.py#L432-L491

- It doesn't support (x :Scalar, alpha: Tensor, beta: Scalar)
                     (x :Scalar, alpha, Scalar, beta: Tensor)

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -7146,6 +7146,402 @@ Tensor values_backward(const Tensor& grad, const Tensor& self) {
+    stop = std::get<0>(ret);
+    values = std::get<2>(ret);
+    gradients = std::get<3>(ret);
+    if (stop.all().item<bool>()) //TODO: It can be bottleneck..
```
Any Suggestion for Performance Improvement?

albanD replied:
On GPU it might be more performant to always do the 100 iterations here? Maybe we could do more fancy async checking of these conditions, but I would definitely keep that for a follow up PR.
For CPU, this will be good for sure.

voidbag commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -7146,6 +7146,402 @@ Tensor values_backward(const Tensor& grad, const Tensor& self) {
+    stop = std::get<0>(ret);
+    values = std::get<1>(ret);
+    gradients = std::get<2>(ret);
+    if (stop.all().item<bool>()) //TODO: It can be bottleneck..
```
Any Suggestion for Performance Improvement?

voidbag reviewed (commented):
Any Suggestion for Performance Improvement in These Control Flows?
(ex. tf.while_loop)

nicola-decao commented:
My name or email hasn't changed! Thanks for asking! :)

On Tue, 30 Jul 2024, 13:17 Taeksang Kim, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> Any Suggestion for Performance Improvement?
> ------------------------------
>
> In torch/csrc/autograd/FunctionsManual.cpp
> <https://github.com/pytorch/pytorch/pull/132135#discussion_r1696862824>:
>
> > +  Tensor initial_convergent = initial_ratio_denominators;
> +  std::vector<Tensor> values = {std::move(initial_ratio_numerators), std::move(initial_ratio_denominators), std::move(initial_convergent)};
> +
> +  Tensor initial_dratio_denominators = at::cat({1.0 - _b, ap1}, -1) * _x / at::square(_x * apb - ap1);
> +  Tensor initial_dratio_numerators = at::zeros_like(initial_dratio_denominators);
> +  Tensor initial_dconvergent = initial_dratio_denominators;
> +  std::vector<Tensor> gradients = {std::move(initial_dratio_numerators), std::move(initial_dratio_denominators), std::move(initial_dconvergent)};
> +
> +  Tensor stop = ~_use_continued_fraction;
> +
> +  for(int32_t i = 0; i < max_iterations; i++) {
> +    std::tuple<Tensor, int32_t, std::vector<Tensor>, std::vector<Tensor>> ret = __continued_fraction_evaluation(stop, i + 1, values, gradients);
> +    stop = std::get<0>(ret);
> +    values = std::get<2>(ret);
> +    gradients = std::get<3>(ret);
> +    if (stop.all().item<bool>()) //TODO: It can be bottleneck..
>
> Any Suggestion for Performance Improvement?
> ------------------------------
>
> In torch/csrc/autograd/FunctionsManual.cpp
> <https://github.com/pytorch/pytorch/pull/132135#discussion_r1696863099>:
>
> > +  Tensor initial_series_sum = 1.0 / safe_a;
> +  std::vector<Tensor> values = {std::move(initial_n), std::move(initial_product), std::move(initial_series_sum)};
> +
> +  Tensor initial_product_grad_b = at::zeros_like(safe_b);
> +  Tensor initial_da = -at::reciprocal(at::square(safe_a));
> +  Tensor initial_db = initial_product_grad_b;
> +  std::vector<Tensor> gradients = {std::move(initial_product_grad_b), std::move(initial_da), std::move(initial_db)};
> +
> +  Tensor stop =  ~use_power_series;
> +
> +  for(int32_t i = 0; i < max_iterations; i++) {
> +    std::tuple<Tensor, std::vector<Tensor>, std::vector<Tensor>> ret = __power_series_evaluation(stop, values, gradients);
> +    stop = std::get<0>(ret);
> +    values = std::get<1>(ret);
> +    gradients = std::get<2>(ret);
> +    if (stop.all().item<bool>()) //TODO: It can be bottleneck..
>
> Any Suggestion for Performance Improvement?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/132135#pullrequestreview-2207523849>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACKA5PH223SD5CX7VWGDWTTZO574HAVCNFSM6AAAAABLWHOROKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDEMBXGUZDGOBUHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

richwomanbtc commented:
My name and email also unchanged.
Thanks for asking

voidbag commented:
> My name and email also unchanged. Thanks for asking

@richwomanbtc
Thanks for reply.
What do you think about closing related issue and PR you opened?

voidbag committed:
special.betainc: lbeta handles large value >= 8

lbeta in betainc has been modified based on tensorflow_probability impl

- Reference: https://github.com/tensorflow/probability/blob/
  b14ae1d79de4a52d834a3ba2dc88f7e5d849e6c7/
  tensorflow_probability/python/math/special.py#L2015-L2053

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

Sharan123 commented:
Hi! What is the status of this PR? I am in a big need for a Beta.cdf method (or practically a way to calc integral of pdf for a<=x<=b).

richwomanbtc commented:
> > My name and email also unchanged. Thanks for asking
> 
> @richwomanbtc Thanks for reply. What do you think about closing related issue and PR you opened?

No problem

voidbag commented:
> Hi! What is the status of this PR? I am in a big need for a Beta.cdf method (or practically a way to calc integral of pdf for a<=x<=b).

@Sharan123 
I think this PR is waiting to be reviewed.
If you want to use the feature immediately, you can use my custom docker image with `docker pull voidbag/pytorch:2.4.0-cuda12.1-cudnn9-devel-wip-betainc-with-bwd-voidbag-v0.1.0`

voidbag commented:
@bdhirsh 
I have no authority to remove the stale label on this PR.
Can I ask you to remove the stale label on it?

Thanks

hchau630 commented:
Would also like this PR to be reviewed as I also have a need for it.

Oceanusity commented:
The cdf of Beta distribution is a important function and widely used. Hope this PR can be reviewed smoothly. Thanks.

voidbag committed:
native/BetaOps: refactor betainc, betaln, beta

- betainc,betaln has been moved to seperate BetaOps files
- betainc kernel is moved to BetaOpsKernel from BinaryOpsKernel
- betaln has been added to native_functions.yaml
- the partial derivative of betaln has been added to derivatives.yaml

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented:
[update]
- specific kernel ops files(BetaOps) are written for beta distribution family.
- BinaryOps, BinaryOpsKernel, BinaryMiscOpsKernels aren't used, currently
- betatincinv (forward, backward) has been implemented

voidbag committed:
betaincc, betainccinv have been added to special

betaincc and betainccinv have been implemented by python code

the code is in torch/special/__init__.py
- betaincc(x): 1 - betainc(x)
- betainccinv(y): betaincinv(1 - y)

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
native/BetaOps: betaincinv have been added

- betaincinv has been implemented
  forward op is in native/BetaOps.cpp
  backward op is in FunctionsManual
- Reference: https://github.com/tensorflow/probability/blob/
  751461f84b8d29c83216d719950b8d999c7381d5/
  tensorflow_probability/python/math/special.py#L567-L868

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
torch/special/__init__.py: docstring has been written

The docstring has been written for the following functions,
betainc, betaincc, betaincinv, betainccinv, betaln and beta.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented:
Discussion Point (The order of parameters  of betainc, betaincinv, betaincc, betainccinv)

ex) betainc
Current Implementation in This PR
- forward: betainc(x, alpha, beta) (from #58700)
- backward: betainc_bwd(alpha, beta, x)

Other betainc Implementations 
- scipy.special.betainc(alpha, beta, x)
- tfp.math.betainc(alpha, beta, x)

If torch should follow the parameter order of other implementations, I'll change the parameter order of forward functions like scipy and tensorflow probability.
Otherwise, I'll change the parameter order of backward functions to be consistent with forward function.

Thanks.

bdhirsh commented:
@albanD @lezcano do you know who would be a good reviewer here?

albanD commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -466,7 +466,7 @@ void TensorIteratorBase::compute_types(const TensorIteratorConfig& config) {
   //   - creates temporaries for CPU operations, if needed and requested
   common_device_ = common_device;
-  int max_cpu_scalars_on_non_cpu = config.allow_cpu_scalars_ ? 1 : 0;
+  int max_cpu_scalars_on_non_cpu = config.allow_cpu_scalars_ ? config.num_inputs_ - 1 : 0;
```
Is there any other change needed down the line to support this?

voidbag replied:
I think max_cpu_scalars_on_non_cpu supports only binary inputs initially.
But, the number of inputs of the gpu_kernel_with_scalars_ternary I wrote is 3.
Exiting code caused errors when executing gpu_kernel_with_scalars_ternary, because of the above constraint (1 or 0). So I made the constraint support the ternary function.

However,
if the tensors wrapping scalars are used, that change might be reverted, I think.

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -13478,6 +13478,173 @@
+  variants: function
+  structured_delegate: special_betainc.out
+
+- func: special_betainc.x_scalar(Scalar self, Tensor a, Tensor b) -> Tensor
```
Do we really need the scalar overloads for these?
We would wrap python numbers into Tensors automatically and that would mean we only need the Tensor/Tensor/Tensor overload above.

voidbag replied:
I wrote the scalar overloads, to correspond with @nicola-decao 's code.
I'll change the scalar overloads to wrapped python numbers, as you suggested.

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -13478,6 +13478,173 @@
+  dispatch:
+    CompositeExplicitAutograd: special_betaln
+
+- func: special_betaln.a_scalar(Scalar a, Tensor b) -> Tensor
```
Similarly here to remove the scalar overloads.

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -13478,6 +13478,173 @@
+  dispatch:
+    CompositeExplicitAutograd: special_betaln_out
+
+- func: special_betaincinv(Tensor self, Tensor a, Tensor b) -> Tensor
```
Same here

albanD commented on torch/csrc/api/include/torch/special.h:
```diff
@@ -452,6 +452,116 @@ inline Tensor& zeta_out(
   return torch::special_zeta_out(result, self, other);
 }
 
+/// Computes betaln
```
I would suggest removing the scalar overloads here as well to simplify significantly the logic here.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -7164,4 +7165,549 @@ Tensor values_backward(const Tensor& grad, const Tensor& self) {
+   * here: https://dlmf.nist.gov/8.17.E23
+   */
+  auto options = at::TensorOptions().dtype(x.dtype()).device(x.device());
+  const Tensor one = at::scalar_tensor(1.0, options);
+  const Tensor two = at::scalar_tensor(2.0, options);
```
wrapping should be done automatically below no? You should be able to use 1. or 2. in the formulas

voidbag replied:
I defined the constant scalar tensors (1., 2., .5, etc) explicitly, to have the same data type and the same device with other operands. The constant tensors can be reused in following formulas.

It's the coding style corresponding with the python codes of tensorflow-probability.
I guess the tf-probability tried to avoid wrapping python number objects dynamically and repeatedly, for performance.
I think that the c++ codes of PyTorch can be different from the python codes of tf-probability.
For example, the tensors wrapping 'rvalue constant scalar values' might be reused automatically in compile time.

If you think the definitions aren't helpful for performance, I'll change the definitions to rvalue constants like 1. or 2.

albanD replied:
We don't really wrap numbers into Tensors using our "Scalar" overloads! So these are just kept as full precision numbers and avoid using Tensors altogether.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -7164,4 +7165,549 @@ Tensor values_backward(const Tensor& grad, const Tensor& self) {
+   *     https://www.fresco.org.uk/papers/Thompson-JCP64p490.pdf
+   */
+  // a, b, and x have same dtype
+  std::tuple<Tensor, Tensor> eps_tiny = AT_DISPATCH_FLOATING_TYPES_AND2(
```
```suggestion
  auto {eps, tiny} = AT_DISPATCH_FLOATING_TYPES_AND2(
```
This should work to simplify the logic below.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -7164,4 +7165,549 @@ Tensor values_backward(const Tensor& grad, const Tensor& self) {
+        x.scalar_type(),
+        "__betainc_modified_lentz_method_eps_tiny",
+        [&]() -> std::tuple<Tensor, Tensor> {
+    Tensor eps = at::scalar_to_tensor(std::numeric_limits<at::scalar_value_type<scalar_t>::type>::epsilon(), x.device());
```
Not sure why you use the AT_DISPATCH_FLOATING_TYPES_AND2 macro here. You can get these values from x.scalar_type() directly no?

voidbag replied:
I referenced the below code snippet in **torch/csrc/TypeInfo.cpp** for getting eps, tiny, max, min, etc of floating type.
```
#define _AT_DISPATCH_FINFO_TYPES(TYPE, NAME, ...) \
  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND6(    \
      at::kHalf,                                  \
      at::ScalarType::BFloat16,                   \
      at::ScalarType::Float8_e5m2,                \
      at::ScalarType::Float8_e5m2fnuz,            \
      at::ScalarType::Float8_e4m3fn,              \
      at::ScalarType::Float8_e4m3fnuz,            \
      TYPE,                                       \
      NAME,                                       \
      __VA_ARGS__)

static PyObject* THPFInfo_eps(THPFInfo* self, void*) {
  HANDLE_TH_ERRORS
  return _AT_DISPATCH_FINFO_TYPES(self->type, "epsilon", [] {
    return PyFloat_FromDouble(
        std::numeric_limits<at::scalar_value_type<scalar_t>::type>::epsilon());
  });
  END_HANDLE_TH_ERRORS
}
```
The scalar type seems to be given to the template  `std::numeric_limits<at::scalar_value_type<scalar_t>::type>::epsilon()` in compile time, to get epsilon value.

I don't know how to get floating numbers' information(eps, tiny, min, max, etc)  in run time.
If you let me know how to get the float information in run time, I'll use it.

voidbag replied:
> On GPU it might be more performant to always do the 100 iterations here? Maybe we could do more fancy async checking of these conditions, but I would definitely keep that for a follow up PR.
For CPU, this will be good for sure.

I think it might be not good to remove control flow with maximum iteration loop.
When I trained model(LBD, recsys23), the train iteration becomes slower and slower.(2~4 times?, not accurate)
In tensorflow-probability, same situation occurs (slower iterations), which means bugs in PyTorch didn't make iterations slower, I think.
I guess the break point becomes longer in the indefinite loop in betainc's bwd as training process is going on in LBD(recsys23) model.


So, I think it's better to do more fancy async checking that you said than to execute maximum iterations statically(loop unrolling?). I think the fancy async checking can be done in follow-up PR, as you said.

albanD reviewed (commented):
Thanks for collecting the PRs and pushing this.
I think there are quite a few big changes that are needed here. Mainly:
- Show the benefit of writing all of these formulas by hand rather than just letting autograd differentiate through the lgamma implementation. If that works, we can define these functions as either python functions directly or CompositeImplicitAutograd formulas that don't need an explicit autograd formula.
- Simplify the code by removing scalar overloads, simplify some code construct in the autograd formulas
- Add testing for both forward and backward through OpInfo (see other ops in common_methods_invocation.py).

voidbag commented:
@albanD
> 1. Show the benefit of writing all of these formulas by hand rather than just letting autograd differentiate through the lgamma implementation. If that works, we can define these functions as either python functions directly or CompositeImplicitAutograd formulas that don't need an explicit autograd formula.
> 2. Simplify the code by removing scalar overloads, simplify some code construct in the autograd formulas
> 3. Add testing for both forward and backward through OpInfo (see other ops in common_methods_invocation.py).

1. the forward and backward of betainc and betaincinv is calculated **approximately**. I Think the autograd differentiation of approximated forward function isn't good. In case of betaln(x,y), the fwd can be calculated using lgamma where x and y < 8. But, in case of other x and y, approximation process is also needed in betaln(x,y). So the three functions(**betainc, betaln, betaincinv**) can't use the advantages of autograd differentiation. That's why the formulas were written by hand in tensorflow-probability, I think. The other three functions (betaincc, betainccinv, beta) can enjoy autograd differentiation. So I defined the other three functions in torch/special/__init__.py.
2. I'll do that based on what you suggested.
3. I'll add testing codes as soon as **2** completes.

voidbag commented:
build error occurs in previous workflow (https://github.com/pytorch/pytorch/actions/runs/12453337435/job/34806491306?pr=132135)

https://github.com/pytorch/pytorch/actions/runs/12453337435/job/34806491306?pr=132135
1. **static** keyword has been added to each static function in FunctionsManual.cpp (resolved)
2. Failure to include library `#include <unsupported/Eigen/CXX11/Tensor`
    ```
    /opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/include/ATen/native/Math.h:16:10: fatal error: unsupported/Eigen/CXX11/Tensor: No such file or directory
       16 | #include <unsupported/Eigen/CXX11/Tensor> // from @eigen_archive
          |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    compilation terminated.
    ```

voidbag committed:
FunctionsManual: refactor betainc,betaincinv

1. remove the scalar overload of backward operations of
   betainc and betaincinv.
2. Use bwd with only tensor parameters, using lambda functions in
   derivatives.yaml.
3. add jvp operation to betainc and betaincinv in derivatives.yaml.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented on tools/autograd/derivatives.yaml:
```diff
@@ -1031,6 +1031,72 @@
+  result: "[&]() { auto [g_a, g_b, g_self] = betainc_partials(a_p, b_p, self_p);
+                   return g_a * a_t + g_b * b_t + g_self * self_t; }()"
+
+- name: special_betainc.x_scalar(Scalar self, Tensor a, Tensor b) -> Tensor
+  a, b: "[&]() { auto [g_a, g_b, g_self] = special_betainc_backward(grad, a, b, at::scalar_to_tensor(self, a.device()));
+                 return std::make_tuple(std::move(g_a), std::move(g_b)); }()"
+  result: "[&]() { auto [g_a, g_b, g_self] = betainc_partials(a_p, b_p, at::scalar_to_tensor(self, a_p.device()));
+                   return g_a * a_t + g_b * b_t; }()"
```
I replaced a lot of verbose **backward** functions with a tensor tensor function receiving wrapped scalar values.

I don't remove the scalar overloads in fwd yet, because It's useful to provide c++ forward functions with scalar overloads.

And I'm not sure of where the code wrapping python number is located. (`torch/special/__init__.py` or `torch/_refs/special/__init__.py`?)
I don't think that `torch/_refs/special/__init__.py` is appropriate, because betainc, betaln and betaincinv cannot be decomposed. Then the other option is `torch/special/__init__.py`. I'm worried about verbose codes in `torch/special/__init__.py` which is a very simple file.

I think it might be okay to leave the scalar overloads for future c++ usage. What about you? @albanD

albanD replied:
> And I'm not sure of where the code wrapping python number is located. 

It is done by our code-generated python bindings. So when you call torch.foo(2.) in python, the python binding will nicely wrap that into a Tensor and at::foo will receive a Tensor.

Let me take another look at the code for the scalar overloads. But my current thinking is:
- C++ usage is relatively limited so I think it's fair to not provide everything there to keep things simpler.
- In particular, new entries here lead to quite a bit of codegen, python bindings and compilation time.
- If we do want to allow these, I think it is fair to add these conversions in the torch/include/* bindings for the C++ API by hand. This will reduce the amount of code/ops that we add while keeping the feature if users need it.

voidbag commented:
WIP
1. refactoring
    - [ ] change constant tensors to rvalue(float)
    - [ ] remove scalar overloads (??)
2. `torch/testing`
    - [x] Composite Compliance
        `stop.all().item<bool>()` -> `stop.all().equal(_true); // where _true = at::scalar_tensor(true, x.device())`
    - [ ] inductor
        `#include <unsupported/Eigen/CXX11/Tensor>` in `Math.h` causes error when test code trying to compile `Math.h` with g++.
        Temporary Resolution: `-I/${TORCH_ROOT}/third_party/nlohmann/`..
    - [ ] betaincinv, betaln
        ```NotImplementedError: aten::equal: attempted to run this operator with Meta tensors, but there was no fake impl or Meta kernel registered. You may have run into this message while using an operator with PT2 compilation APIs (torch.compile/torch.export); in order to use this operator with those APIs you'll need to add a fake impl. Please see the following for next steps:  https://pytorch.org/tutorials/advanced/custom_ops_landing_page.html```

voidbag committed:
common_methods_invocations: add ops of beta family

The functions added in common_methods_invocations.py are
special.betainc, special.betaincinv and special.betaln.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
BetaOps: remove scalar overloads

1. Removed scalar overloads of betainc, betaincinv and betaln
2. add betainc, betaincinv and betaln into allowed set in
   should_allow_numbers_as_tensors (python_args_parser.cpp), to wrap
   python number objects into tensors automatically.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -97,6 +97,9 @@ bool should_allow_numbers_as_tensors(const std::string& name) {
       "true_divide",  "true_divide_",  "true_divide_out",
       "to",           "_to_copy",      "copy_",
       "floor_divide", "floor_divide_", "floor_divide_out",
+      "special_betainc", "special_betainc_out",
+      "special_betaincinv", "special_betaincinv_out",
+      "special_betaln", "special_betaln_out",
```
To wrap python number object into tensor, native beta family functions have been added to allowed set.
Is it okay to add these entries here? @albanD

albanD replied:
Yes that's the right place to put them!
The PR is a lot smaller now, thanks!

voidbag replied:
lintrunner fails due to code format.
To pass the check, the codes should be changed as below, including many new lines.
Is it okay to insert some new lines for passing lintrunner?

```
>>> Lint for torch/csrc/utils/python_arg_parser.cpp:

  Warning (CLANGFORMAT) format
    See https://clang.llvm.org/docs/ClangFormat.html.
    Run `lintrunner -a` to apply this patch.

    You can run `lintrunner -a` to apply this patch.

     87   87 | // functions.)
     88   88 | bool should_allow_numbers_as_tensors(const std::string& name) {
     89   89 |   static std::unordered_set<std::string> allowed = {
     89      |-      "add","add_","add_out",
     90      |-      "div","div_","div_out",
     91      |-      "divide","divide_","divide_out", // alias of div
     92      |-      "mul","mul_","mul_out",
     93      |-      "multiply","multiply_","multiply_out", // alias of mul
     94      |-      "sub","sub_","sub_out",
     95      |-      "subtract","subtract_","subtract_out", // alias of sub
     96      |-      "true_divide","true_divide_","true_divide_out",
     97      |-      "to","_to_copy",      "copy_",
     98      |-      "floor_divide","floor_divide_","floor_divide_out",
     99      |-      "special_betainc","special_betainc_out",
    100      |-      "special_betaincinv","special_betaincinv_out",
    101      |-      "special_betaln","special_betaln_out",
          90 |+      "add",
          91 |+"add_",
          92 |+"add_out",
          93 |+      "div",
          94 |+"div_",
          95 |+"div_out",
          96 |+      "divide",
          97 |+"divide_",
          98 |+"divide_out", // alias of div
          99 |+      "mul",
         100 |+"mul_",
         101 |+"mul_out",
         102 |+      "multiply",
         103 |+"multiply_",
         104 |+"multiply_out", // alias of mul
         105 |+      "sub",
         106 |+"sub_",
         107 |+"sub_out",
         108 |+      "subtract",
         109 |+"subtract_",
         110 |+"subtract_out", // alias of sub
         111 |+      "true_divide",
         112 |+"true_divide_",
         113 |+"true_divide_out",
         114 |+      "to",
         115 |+"_to_copy",
         116 |+      "copy_",
         117 |+      "floor_divide",
         118 |+"floor_divide_",
         119 |+"floor_divide_out",
         120 |+      "special_betainc",
         121 |+"special_betainc_out",
         122 |+      "special_betaincinv",
         123 |+"special_betaincinv_out",
Traceback (most recent call last):
  File "/home/ec2-user/actions-runner/_work/pytorch/pytorch/test-infra/.github/scripts/run_with_env_secrets.py", line 102, in <module>
         124 |+      "special_betaln",
         125 |+"special_betaln_out",
    103  126 |       "_conj"}; // _conj needed because mul.Tensor backward calls it
    104  127 |   return allowed.find(name) != allowed.end();
    105  128 | }
```

albanD replied:
Yes, whatever lintrunner wants, lintrunner gets :p

voidbag committed:
BetaOps: scalar tensor has been removed.

Scalar tensors have been replaced with cpp rvalue constant.
- The tensors used as tensor arguments in function weren't removed.
- The replacement that generates operations btw only cpp constants isn't
  permitted.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented:
@albanD 
I modified codes based on your suggestion.
Could you review this PR again?

I have a another question.
`#include <unsupported/Eigen/CXX11/Tensor>` in Math.h causes dynamo test failure. (g++ cannot find the directory).
I don't know how to include the third_party directory(third_party/eigen) in that test.
Could you please help me?

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -12028,6 +12093,39 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
                DecorateInfo(unittest.skip("Test expects tensor input"), "TestVmapOperatorsOpInfo", "test_vmap_exhaustive"),
                DecorateInfo(unittest.skip("Test expects tensor input"), "TestVmapOperatorsOpInfo", "test_op_has_batch_rule"),
            )),
+    OpInfo('special.betainc',
```
Can you add a numpy-based reference functions using `ref=` to ensure the value we compute is good?

voidbag replied:
I added `ref=` for checking functions' correctness

albanD reviewed (commented):
The PR looks good, thanks for the hard work!
I haven't gone through the kernel in details yet but adding the ref on top of the opinfo tests you already have should make sure the results are good!

> #include <unsupported/Eigen/CXX11/Tensor> in Math.h causes dynamo test failure. (g++ cannot find the directory).

Not sure! I triggered CI here and we'll see what fails exactly!

voidbag committed:
Loops.cuh: remove gpu_kernel_with_scalars_ternary

gpu_kernel_with_scalars_ternary hasn't been used anymore.
For future implementation, I left the commit removing the function.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented:
Added Commits: 
Loops.cuh: remove gpu_kernel_with_scalars_ternary

The feature "ternary gpu kernel including scalars" has been removed, because scalar overloads weren't used any more in beta family functions.
But, I think 'gpu_kernel_with_scalars_ternary' could be used for future PR. That's why I left the commit removing the function.

voidbag committed:
BetaOps: change params' order and add scipy ref

- Change the order of parameter from (x, a, b) -> (a, b, x)
- Add ref functions 'scipy.special.(betainc|betaincinv|betaln)' in
  common_methods_invocations.py

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented:
There are big changes.

The order of parameters of betainc, betaincc, betaincinv and betainccinv has been changed from (x, a, b) to (a, b, x) which is consistent with scipy.special, tfp.math module.

voidbag commented:
> The PR looks good, thanks for the hard work! I haven't gone through the kernel in details yet but adding the ref on top of the opinfo tests you already have should make sure the results are good!
> 
> > #include <unsupported/Eigen/CXX11/Tensor> in Math.h causes dynamo test failure. (g++ cannot find the directory).
> 
> Not sure! I triggered CI here and we'll see what fails exactly!

The failed CI log is https://github.com/pytorch/pytorch/actions/runs/12581669412

albanD commented:
Thanks for the update, having the same signature as the scipy functions is definitely what we want!

albanD commented:
Github had some issues yesterday which made CI workflows fail for no reason. Re-running again now.

albanD reviewed (commented):
I see the issue. Not super clear indeed.
@eellison does this ring a bell that the inductor-generated cpp files don't actually have eigen as a dependency and so any code from aten being used there cannot use eigen?

@voidbag I think a simple workaround for you here in case Elias doesn't have a simple answer for us is to move the definitions (and include) you added to ATen/native/Math.h to another file that is not included by inductor codegen, the newly added ATen/native/BetaOps.h looks like a good candidate.
Could you try that?

voidbag commented:
> I see the issue. Not super clear indeed. @eellison does this ring a bell that the inductor-generated cpp files don't actually have eigen as a dependency and so any code from aten being used there cannot use eigen?
> 
> @voidbag I think a simple workaround for you here in case Elias doesn't have a simple answer for us is to move the definitions (and include) you added to ATen/native/Math.h to another file that is not included by inductor codegen, the newly added ATen/native/BetaOps.h looks like a good candidate. Could you try that?

I'll try to do that!

voidbag committed:
BetaOps.h: add betainc_impl from eigen

- add codes about betainc to BetaOps.h based on
  `third_party/eigen/unsupported/Eigen/src/SpecialFunctions/SpecialFunctionsImpl.h`
- remove the codes including eigen repo to resolve CI issue.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented on aten/src/ATen/native/BetaOps.h:
```diff
@@ -0,0 +1,528 @@
+    }
+
+    /* Multiply w by the factor
+         a      b   _             _     _
+        x  (1-x)   | (a+b) / ( a | (a) | (b) ) .   */
+
+    y = a * std::log(x);
+    t = b * std::log(xc);
+    // TODO: gamma is not directly implemented in Eigen.
+    /*
+    if ((a + b) < maxgam && numext::abs(y) < maxlog && numext::abs(t) < maxlog)
+    {
+      t = pow(xc, b);
+      t *= pow(x, a);
+      t /= a;
+      t *= w;
+      t *= gamma(a + b) / (gamma(a) * gamma(b));
+    } else {
+    */
+    /* Resort to logarithms.  */
+    y += t + std::lgamma(a + b) - std::lgamma(a) -
+         std::lgamma(b);
+    y += std::log(w / a);
+    t = std::exp(y);
+
+    /* } */
+    // done:
+
+    if (reversed_a_b) {
+      if (t <= machep) {
+        t = 1.0 - machep;
+      } else {
+        t = 1.0 - t;
+      }
+    }
+    return t;
+  }
+};
+
+/* end of third_party/eigen/unsupported/Eigen/src/SpecialFunctions/SpecialFunctionsImpl.h */
```
@albanD 
The definition of betainc has been added to BetaOps.h from `third_party/eigen/unsupported/Eigen/src/SpecialFunctions/SpecialFunctionsImpl.h` as you suggested!

voidbag commented:
Oh, I made a mistake for changing code signature in 
```
def cdf(self, x):
        return torch.special.betainc(x, self.concentration1, self.concentration0) 
```
`torch.special.betainc(x, self.concentration1, self.concentration0)` must be changed to `torch.special.betainc(self.concentration1, self.concentration0, x)`

voidbag committed:
torch/special/__init__.py: add funcs to __all__

- add beta functions in special module to __all__ in '__init__.py'

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented:
@albanD 
Would you trigger the CI?

eellison commented:
@albanD I would imagine there is a way to conditionally add deps. But maybe adding it in Math is better. cc @desertfire who is a little more knowledgeable on cpp codegen

voidbag commented:
I modified the code using `lintrunner -a`

albanD commented:
All the PT2 related failures are because we don't handle the _special namespace properly there. While I'm looking into a better fix, for this PR, you can just add the new functions you added here to https://github.com/pytorch/pytorch/blob/0f051eaf66b5045fe0131e75966be2cb7645687b/torch/_dynamo/trace_rules.py#L1184 next to the others!

yanboliang commented:
> FAILED [0.0195s] inductor/test_torchinductor_codegen_dynamic_shapes.py::DynamicShapesCodegenCpuTests::test_pointwise_beta_dynamic_shapes_cpu - torch._dynamo.exc.Unsupported: 'skip function beta in file /opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/special/__init__.py'

You have to add the list of functions you added in this PR to `trace_rules.py` following this line:
https://github.com/pytorch/pytorch/blob/d7f45fc575fc77fa4829038fda9cad8a90b79c64/torch/_dynamo/trace_rules.py#L1239 
It means these functions would be treated as `allow in graph` function during Dynamo tracing.

voidbag commented:
To Do List
- [ ]  **Error**: `AssertionError: The following functions are not tested for __torch_function__ support, please ensure there is an entry in the dict returned by torch.overrides.get_testing_overrides for this function or if a __torch_function__ override does not make sense, add an entry to the tuple returned by torch._overrides.get_ignored_functions.`
  **shell**:  `python -bb test/test_overrides.py --shard-id=1 --num-shards=1 -v -vv -rfEX -p no:xdist --use-pytest -x --reruns=2 --import-slow-tests --import-disabled-tests`
 - [ ] **Error**: `AssertionError: False is not true : All the APIs below do not meet our guidelines for public API from https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation.
Make sure that everything that is public is expected (in particular that the module has a properly populated `__all__` attribute) and that everything that is supposed to be public does look public (it does not start with `_` and has a `__module__` that is properly populated).`
  **Shell**: `python test/test_public_bindings.py TestPublicBindings.test_correct_module_names`
- [ ]  **Error**: 
  **Shell**:
- [ ]  **Error**: 
  **Shell**
- [ ]  **Error**: 
  **Shell**
- [ ]  **Error**: 
  **Shell**
- [ ]  **Error**: 
  **Shell**
- [ ]  **Error**: 
  **Shell**

voidbag commented:
@albanD 
I summarized the errors to be resolved with example test command.
I tried to do the example command in local environments. But I faced an issue.

Test 3, 4, 5, 7 and 10 have the below exception about torchvision in my build environment. (RuntimeError: operator torchvision::nms does not exist)
How can I handle the torchvision issue?

I'll try to handle other issues first!
Thanks

Example)
3 Test: `PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=0 python test/functorch/test_aotdispatch.py TestEagerFusionOpInfoCPU.test_aot_autograd_exhaustive_special_betaincinv_cpu_float32 PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=0 PYTORCH_TEST_WITH_ASAN=1 PYTORCH_TEST_WITH_UBSAN=1 python test/functorch/test_aotdispatch.py`
```
Traceback (most recent call last):
  File "/data/workspace/special.betainc/origin/pytorch/test/functorch/test_aotdispatch.py", line 92, in <module>
    import torchvision
  File "/opt/conda/lib/python3.10/site-packages/torchvision/__init__.py", line 10, in <module>
    from torchvision import _meta_registrations, datasets, io, models, ops, transforms, utils  # usort:skip
  File "/opt/conda/lib/python3.10/site-packages/torchvision/_meta_registrations.py", line 164, in <module>
    def meta_nms(dets, scores, iou_threshold):
  File "/opt/conda/lib/python3.10/site-packages/torch/library.py", line 854, in register
    use_lib._register_fake(op_name, func, _stacklevel=stacklevel + 1)
  File "/opt/conda/lib/python3.10/site-packages/torch/library.py", line 198, in _register_fake
    handle = entry.fake_impl.register(func_to_register, source)
  File "/opt/conda/lib/python3.10/site-packages/torch/_library/fake_impl.py", line 31, in register
    if torch._C._dispatch_has_kernel_for_dispatch_key(self.qualname, "Meta"):
RuntimeError: operator torchvision::nms does not exist

```

albanD commented:
Note that I expect most of the CI failures to be due to the trace_rules.py missing change mentioned above.
I would do this one before checking the other failures !

voidbag committed:
Handle PT2 related failures

- add entries to `torch/_dynamo/trace_rules.py`

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented:
I added entries in trace_rules.py first.

voidbag committed:
BetaOps: make special constants have correct dtype

- make special constants(eps, tiny, min, nan) have correct dtype

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -7164,4 +7165,574 @@ Tensor values_backward(const Tensor& grad, const Tensor& self) {
+
+  Tensor stop = ~_use_continued_fraction;
+
+  for (int32_t i = 0; i < max_iterations; i++) {
+    std::tuple<Tensor, int32_t, std::vector<Tensor>, std::vector<Tensor>> ret =
+        __continued_fraction_evaluation(stop, i + 1, values, gradients);
+    stop = std::get<0>(ret);
+    values = std::get<2>(ret);
+    gradients = std::get<3>(ret);
+    if (stop.all().equal(_true)) // TODO: It can be bottleneck..
+      break;
+  }
```
I'm handling test 3, first.

When executing `PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=0 python test/functorch/test_aotdispatch.py TestEagerFusionOpInfoCPU.test_aot_autograd_exhaustive_special_betaincinv_cpu_float32`,

The code line **`stop.all().equal(_true)`** on a data-dependent control flow(indefinite loop) in `FunctionsManual.cpp` causes  `File "/opt/conda/lib/python3.11/site-packages/torch/_subclasses/fake_impls.py", line 554, in data_dep raise DataDependentOutputException(func) torch._subclasses.fake_tensor.DataDependentOutputException: aten.equal.default`

### Is there any way to make the data dependent code block **"one big atomic operations"** in aot_inductor's view ?

Even if removing the if statement makes test passed, I cannot remove the if statement for performance issue.

I already read https://github.com/pytorch/pytorch/blob/b5655d9821b7214af200d0b8796a10ad34b85229/functorch/docs/source/ux_limitations.rst#data-dependent-python-control-flow.

But, I don't know if it's possible to handle **c++**(not python) data dependent control flow.

voidbag replied:
https://github.com/pytorch/pytorch/blob/922f53930803f08c8ac8616d6f4673be2e3792b1/torch/csrc/autograd/FunctionsManual.cpp#L6939-L6941

The workaround for this issue is to change `if (stop.all().equal(_true))` into `if (at::GradMode::is_enabled() && stop.all().equal(_true)) ` like #106429.
Ironically, It can cause performance degradation in PT2.

albanD replied:
Moving to a native function is ok.
You can also skip the this particular test as expected failure if you need !

voidbag replied:
Oh, I didn't push aten native backward, yet.
I'm implementing another workaround..

If I finish the other workaround, then I'll request CI!

voidbag committed:
BetaOps: make native backwards betainc, betaincinv

Move backward related functions from FunctionsManual.cpp to BetaOps.cpp
It resolves the issues about test_aotdispatch.py

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
Math: move calc_betainc from BetaOps.h to Math.h

This migration is for acessing calc_betainc in
torch/_inductor/codegen/common.py

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
fallback_ops: add beta family fwd, bwd to fallback

To use native fwd and bwd ops containing at::equal in inductor,
   the ops are converted to fallback_ops.

Changing ops with at::equal to fallback resolves issues in
test_torchinductor_opinfo.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
Resolve some test cases

- resolve test_public_bindings
- resolve test_decomp cases
- skip vmap-related cases

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented:
@albanD 
Long time no see!
I returned from my vacation.

I handled inductor's test cases, using fallback and native backward functions.
Additionally, I skipped some tests related to vmap.

Could you please run CI?

voidbag commented:
@albanD 
I leave just a friendly reminder for the CI request.
Please check the request at your convenience.

voidbag commented:
It's just rebased for handling docker image errors in CI.

albanD commented on torchgen/static_runtime/generator.py:
```diff
@@ -203,6 +203,11 @@ def has_alias(
         "alias_copy",
         "_triton_multi_head_attention",
         "special_airy_ai",
+        "special_betainc",
```
I don't think static_runtime is used anymore. Let's keep these functions out of this to not change existing systems.

albanD reviewed (commented):
Only a couple things needed before we can merge.
I think we can improve the implementations a little, but let's do that in a follow up!

voidbag commented:
@albanD 
Could you please reopen this PR that the bot closed automatically?

voidbag commented:
@albanD
Could you please reopen this PR that the bot closed automatically?

I'll update this branch based on your opinion as soon as possible in 30 days!

And
I'll do rebase on main branch to prevent the bot from closing this PR automatically in 30 days, if there is no changes in this branch.

voidbag commented:
@albanD
I know you're super busy.
It would be greatly appreciated if you could reopen this PR!

voidbag committed:
Remove unnecessary code based on reveiw(albanD)

Removed unnecessary codes.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

nealockwood commented:
thank you so much @voidbag for your fastidious work. Just want to echo earlier comments and say this would be a very meaningful addition

voidbag committed:
BetaOps: bwd aten fallback to avoid graph break

Remove composite key from bwd ops to avoid graph break when
torch.compile is executed.

Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag commented:
Hi @albanD,

I’m finally back to update this PR, which has been open since July 2024. I sincerely apologize for the long delay.
In this update, I’ve refactored the dispatch logic to support torch.compile. Specifically, I removed the Composite keys for the backward operators and registered them directly under CPU and CUDA dispatch keys. This is a deliberate choice to ensure a stable ATen fallback and avoid graph breaks caused by the data-dependent (indefinite) loops within its numerical kernels.

I would greatly appreciate it if you could take another look and trigger the CI for testing. Thank you so much for your patience and for maintaining this PR for so long!
